File: kennardstonealgorithm.m
function [ selectedsamplenumbers, remainingsamplenumbers] = kennardstonealgorithm( X, k)
% select samples using Kennard-Stone algorithm
%
% --- input ---
% X : dataset of X-variables (samples x variables)
% k : number of samples to be selected
%
% --- output ---
% selectedsamplenumbers : selected sample numbers (training data)
% remainingsamplenumbers : remaining sample numbers (test data)
% 
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

selectedsamplenumbers = zeros( 1, k);
remainingsamplenumbers = 1 : size( X, 1);
originalX = X;

[~, selectedsamplenumbers(1) ] = max(pdist2(X, mean(X)));
X( selectedsamplenumbers(1), :) = [];
remainingsamplenumbers(selectedsamplenumbers(1)) = [];

for iteration = 1 : k-1
    if iteration == 1
        [~, selectedsamplenumber] = max( pdist2( originalX(selectedsamplenumbers(1:iteration),:), X) );
    else
        [~, selectedsamplenumber] = max( min( pdist2( originalX(selectedsamplenumbers(1:iteration),:), X) ) );
    end
	selectedsamplenumbers(iteration+1) = remainingsamplenumbers(selectedsamplenumber);
    X( selectedsamplenumber, :) = [];
    remainingsamplenumbers(selectedsamplenumber) = [];
end

File: kennardstonealgorithm.py
# select samples using Kennard-Stone algorithm
import numpy as np

# --- input ---
# X : dataset of X-variables (samples x variables)
# k : number of samples to be selected
#
# --- output ---
# selectedsamplenumbers : selected sample numbers (training data)
# remainingsamplenumbers : remaining sample numbers (test data)

def kennardstonealgorithm( X, k ):
    X = np.array( X )
    originalX = X
    distancetoaverage = ( (X - np.tile(X.mean(axis=0), (X.shape[0], 1) ) )**2 ).sum(axis=1)
    maxdistancesamplenumber = np.where( distancetoaverage == np.max(distancetoaverage) )
    maxdistancesamplenumber = maxdistancesamplenumber[0][0]
    selectedsamplenumbers = list()
    selectedsamplenumbers.append(maxdistancesamplenumber)
    remainingsamplenumbers = np.arange( 0, X.shape[0], 1)
    X = np.delete( X, selectedsamplenumbers, 0)
    remainingsamplenumbers = np.delete( remainingsamplenumbers, selectedsamplenumbers, 0)
    for iteration in range(1, k):
        selectedsamples = originalX[selectedsamplenumbers,:]
        mindistancetoselectedsamples = list()
        for mindistancecalculationnumber in range( 0, X.shape[0]):
            distancetoselectedsamples = ( (selectedsamples - np.tile(X[mindistancecalculationnumber,:], (selectedsamples.shape[0], 1)) )**2 ).sum(axis=1)
            mindistancetoselectedsamples.append( np.min(distancetoselectedsamples) )
        maxdistancesamplenumber = np.where( mindistancetoselectedsamples == np.max(mindistancetoselectedsamples) )
        maxdistancesamplenumber = maxdistancesamplenumber[0][0]
        selectedsamplenumbers.append(remainingsamplenumbers[maxdistancesamplenumber])
        X = np.delete( X, maxdistancesamplenumber, 0)
        remainingsamplenumbers = np.delete( remainingsamplenumbers, maxdistancesamplenumber, 0)

    return selectedsamplenumbers, remainingsamplenumbers
